### PR TITLE
feat: kyvejs runtime containerization and enhancements

### PR DESCRIPTION
**Pull request by kchrist-rocketfueldev**
Revamp the KYVE protocol node architecture, which currently consists of two main components: the core and the runtime. The primary focus of this overhaul is the enhancement of the runtime component, which, as of now, is exclusively developed by the KYVE team and implemented in TypeScript.

The primary goal is to introduce a standardized Docker interface, opening the door for the broader community to create their own runtimes in their preferred programming languages, as long as they adhere to the specified Docker interface standards. Consequently, the core component will communicate with the runtime via this Docker interface, simplifying the transfer of data between these two vital parts.

This pull request includes the following key updates:
- Definition of gRPC protobuf file.
- Porting of the Tendermint runtime into a standalone Docker container.
- Development of a command-line tool enabling users to interact with core functions through the command line interface.

These changes collectively aim to make the KYVE protocol node more versatile, allowing a wider community of developers to contribute to and extend its functionality.

Resolves  https://linear.app/kyve/issue/TECH-1077
Resolves https://linear.app/kyve/issue/TECH-1076
Resolves https://linear.app/kyve/issue/TECH-1034

### CODE REVIEW

**Comment by troykessler** [comment]
Here my thoughts:

- move `runtime.proto` from common/types to common/protocol since it is the "source of truth" of the runtime <> core communication and is definitely a core feature. Currently the source of truth for the runtime interface is the `runtime.interface.ts`: question here: what will be the truth of source in the future, do we compile the TS interface out of the proto file?
- validateDataItem changed, instead of a boolean it returns a number now (indicating either valid, invalid or abstain), this has to be adjusted so @kyvejs/docker successfully builds again

File: package.json [hunk 1]
```diff
@@ -35,7 +41,9 @@
     "clean": "lerna run clean",
     "lint": "lerna run lint",
     "lint:fix": "lerna run lint:fix",
-    "custom-version": "yarn lerna version --conventional-commits=false"
+    "custom-version": "yarn lerna version --conventional-commits=false",
+    "export-grpc-stubs": "protoc --proto_path=./common/protocol/src/types --plugin=./node_modules/.bin/protoc-gen-ts_proto --ts_proto_opt=snakeToCamel=false --ts_proto_opt=outputServices=grpc-js --ts_proto_out=./common/docker/src common/protocol/src/types/protos/runtime.proto"
   },
+
```
Comment (harrytou) [inline package.json:47]: @kchrist-rocketfueldev Remove this line.

File: package.json [hunk 2]
```diff
@@ -35,7 +41,9 @@
     "clean": "lerna run clean",
     "lint": "lerna run lint",
     "lint:fix": "lerna run lint:fix",
-    "custom-version": "yarn lerna version --conventional-commits=false"
+    "custom-version": "yarn lerna version --conventional-commits=false",
+    "export-grpc-stubs": "protoc --proto_path=./common/protocol/src/types --plugin=./node_modules/.bin/protoc-gen-ts_proto --ts_proto_opt=snakeToCamel=false --ts_proto_opt=outputServices=grpc-js --ts_proto_out=./common/docker/src common/protocol/src/types/protos/runtime.proto"
```
Comment (harrytou) [inline package.json:45]: @kchrist-rocketfueldev I think this should be integrated in a `scripts` directory inside `common`. We had discussed that devs would run the script to generate new gRPC stubs if the `.proto` file has changed before starting/resuming their work.

This should be able to work for any language but right now make it only for TS until we develop the other runtimes.

File: package.json [hunk 3]
```diff
@@ -7,23 +7,29 @@
...
     "nx": "^14.4.3",
     "prettier": "^2.3.0",
-    "rimraf": "^3.0.2"
+    "rimraf": "^3.0.2",
+    "protobufjs": "^6.8.6",
+    "ts-proto": "^1.159.2"
   },
   "scripts": {
     "setup": "lerna clean && yarn install && lerna run build",
+    "docker-setup": "yarn nx reset && yarn install && lerna run build",
```
Comment (harrytou) [inline package.json:32]: @kchrist-rocketfueldev Can we rename this to `build:docker` to follow the same naming convention where `build` is the namespace (or pipeline stage) and `docker` is the target which means you build the docker image.

It follows the same practice as `build:binaries`.

**Review by harrytou — CHANGES_REQUESTED** [review]

File: package.json (hunk 2 above)
Comment (harrytou) [reply package.json:45]: Also the `export-grpc-stubs` script that does not reference the correct path of the `runtime.proto`.

Comment (kchrist-rocketfueldev) [reply package.json]: added a separate script for export grpc stubs for TS. The script will be updated in the future to support other languages

**Review by harrytou — APPROVED** [review]

**Pull request closed without merging**